mainloop.cpp

```diff
@@ -384,12 +384,10 @@ bool Mainloop::add_endpoints(Mainloop &mainloop, struct opt *opt)
                 continue;
             }
 
-            if (_add_tcp_endpoint(tcp.get()) < 0) {
+            if (_add_tcp_endpoint(tcp.release()) < 0) {
```

Comment:
this means you transferred ownership to _add_tcp_endpoint() and it is the new owner of TcpEndpoint. Which includes freeing it in the error path, which you are not doing.